// douban/collection.js
var holder = {
    musicStatusName : function(status) {
        switch (status) {
            case 'listened' :
                return '听过';
            case 'listening' :
                return '在听';
            case 'wish' :
                return '想听';
        }
    },

    bookStatusName : function(status) {
        switch (status) {
            case 'wish' :
                return '想读';
            case 'reading' :
                return '读过';
            case 'read' :
                return '在读';
        }
    },

    movieStatusName : function(status) {
        switch (status) {
            case 'wish' :
                return '想看';
            case 'watched' :
                return '看过';
        }
    },

    determineCat : function(term) {
        return term.substring(term.lastIndexOf('#') + 1);
    },

    processTemplate : function(entry) {
        var dbSubject = entry['db:subject'];
        var term = dbSubject['category']['@term'];
        var cat = holder.determineCat(term);
        return holder[cat + 'StatusName'](entry['db:status']['$t'])
                + dbSubject.title.$t + ' ' + holder.parseLink(dbSubject['link']);
    },

    parseLink : function(links) {
        for (var i =0, l = links.length; i < l; i++) {
            if (links[i]['@rel'] == 'alternate') {
                return links[i]['@href'];
            }
        }
    }
};


module.exports.handleEntry = function(entry) {
    return '[Douban] 我' + holder.processTemplate(entry);
};
